File: modules/widgets/lib/rin/widgets/containers/WCList.cpp
﻿#include "rin/widgets/containers/WCList.hpp"
#include "rin/widgets/ContainerWidgetSlot.hpp"

WCList::WCList(const WidgetAxis& axis) : ContainerWidget()
{
    _axis = axis;
}

WidgetAxis WCList::GetAxis() const
{
    return _axis;
}

void WCList::SetAxis(const WidgetAxis& axis)
{
    auto oldAxis = _axis;
    _axis = axis;
    if (oldAxis != axis)
    {
        CheckSize();
    }
}

void WCList::ArrangeSlots(const Vec2<float>& drawSize)
{
    switch (GetAxis())
    {
    case WidgetAxis::Horizontal:
        {
            Vec2 offset{0.0f};

            for (auto& containerSlot : GetSlots())
            {
                auto widget = containerSlot->GetWidget();
                widget->SetOffset(offset);
                auto widgetDrawSize = widget->GetDesiredSize();
                widgetDrawSize.y = drawSize.y;
                widget->SetSize(widgetDrawSize);
                offset = offset + Vec2{widgetDrawSize.x, 0.0f};
            }
        }
        break;
    case WidgetAxis::Vertical:
        {
            Vec2 offset{0.0f};

            for (auto& containerSlot : GetSlots())
            {
                auto widget = containerSlot->GetWidget();
                widget->SetOffset(offset);
                auto widgetDrawSize = widget->GetDesiredSize();
                widgetDrawSize.x = drawSize.x;
                widget->SetSize(widgetDrawSize);
                offset = offset + Vec2{0.0f, widgetDrawSize.y};
            }
        }
        break;
    }
}

Vec2<float> WCList::ComputeContentSize()
{
    switch (GetAxis())
    {
    case WidgetAxis::Horizontal:
        {
            Vec2 result{0.0f};

            for (auto& containerSlot : GetSlots())
            {
                auto size = containerSlot->GetWidget()->GetDesiredSize();
                result.x += size.x;
                result.y = std::max(result.y, size.y);
            }

            return result;
        }
        break;
    case WidgetAxis::Vertical:
        {
            Vec2 result{0.0f};

            for (auto& containerSlot : GetSlots())
            {
                auto size = containerSlot->GetWidget()->GetDesiredSize();
                result.y += size.y;
                result.x = std::max(result.x, size.x);
            }

            return result;
        }
        break;
    }
    return Vec2{0.0f};
}

File: aerox.Runtime.Widgets/Containers/WCList.cs
using aerox.Runtime.Math;

namespace aerox.Runtime.Widgets.Containers;

/// <summary>
///     A container that draws children left to right
/// </summary>
public class WCList(IEnumerable<Widget> children) : Container(children)
{
    public enum Axis
    {
        Vertical,
        Horizontal
    }

    private Axis _direction = Axis.Vertical;

    public Axis Direction
    {
        get => _direction;
        set
        {
            _direction = value;
            OnDirectionChanged();
        }
    }

    protected virtual void OnDirectionChanged()
    {
        TryUpdateDesiredSize();
    }

    protected override Size2d ComputeDesiredContentSize()
    {
        return _direction switch
        {
            Axis.Horizontal => GetSlots().Aggregate(new Size2d(), (size, slot) =>
            {
                var slotSize = slot.GetWidget().GetDesiredSize();
                size.Width += slotSize.Width;
                size.Height = System.Math.Max(size.Height, slotSize.Height);
                return size;
            }),
            Axis.Vertical => GetSlots().Aggregate(new Size2d(), (size, slot) =>
            {
                var slotSize = slot.GetWidget().GetDesiredSize();
                size.Height += slotSize.Height;
                size.Width = System.Math.Max(size.Width, slotSize.Width);
                return size;
            }),
            _ => throw new ArgumentOutOfRangeException()
        };
    }


    protected override void ArrangeSlots(Size2d drawSize)
    {
        var offset = new Vector2<float>(0.0f);
        switch (_direction)
        {
            case Axis.Horizontal:
            {
                foreach (var slot in GetSlots())
                {
                    var widget = slot.GetWidget();
                    var widgetSize = widget.GetDesiredSize();
                    widget.SetOffset(offset.Clone());
                    widget.SetSize(new Size2d
                    {
                        Width = widgetSize.Width,
                        Height = widgetSize.Height
                    });
                    offset.X += widgetSize.Width;
                }
                break;
            }
            case Axis.Vertical:
                
                {
                    foreach (var slot in GetSlots())
                    {
                        var widget = slot.GetWidget();
                        widget.SetOffset(offset.Clone());
                        var widgetSize = widget.GetDesiredSize();
                        widget.SetSize(new Size2d
                        {
                            Height = widgetSize.Height,
                            Width = widgetSize.Width
                        });
                        offset.Y += widgetSize.Height;
                    }
                }
                break;
            default:
                throw new ArgumentOutOfRangeException();
        }
    }

    public override void OnSlotUpdated(Slot slot)
    {
        TryUpdateDesiredSize();
        ArrangeSlots(GetContentSize());
    }
}